src/watch_state.rs

```diff
@@ -0,0 +1,246 @@
+
+use watchexec::action::Outcome;
+
+/// Tracks the current state of spin watch. Handles transitioning states and converting effects
+/// into outcomes based on the current state.
+#[derive(Debug, Clone)]
+pub struct StateMachine {
+    state: Arc<Mutex<State>>,
+    clear: bool,
+    skip_build: bool,
+}
+
+impl StateMachine {
+    pub fn new(skip_build: bool, clear: bool) -> Self {
+        Self {
+            state: Arc::new(Mutex::new(match skip_build {
+                false => State::Building,
+                true => State::Running,
+            })),
+            clear,
+            skip_build,
+        }
+    }
+
+    /// Get the current state of the state machine.
+    pub fn get_state(&self) -> State {
+        *self.state.lock().unwrap()
+    }
+
+    /// Based on the given effect return the correct outcome. Transition the internal state of the
+    /// machine if necessary.
+    pub fn handle(&self, effect: Effect) -> Outcome {
+        let mut state = self.state.lock().unwrap();
+        tracing::debug!("handling effect {:?} in current state {:?}", effect, *state);
+        let outcome = match *state {
+            State::Building => match effect {
+                Effect::Exit => {
+                    Outcome::if_running(Outcome::both(Outcome::Stop, Outcome::Exit), Outcome::Exit)
+                }
+                Effect::ProcessFailed => Outcome::if_running(Outcome::Stop, Outcome::DoNothing),
+                Effect::ProcessCompleted => {
+                    *state = State::Running;
+                    self.restart_outcome(true)
+                }
+                Effect::ManifestChange => self.restart_outcome(false),
+                Effect::SourceChange => self.restart_outcome(false),
+                Effect::ArtifactChange => Outcome::DoNothing,
+                Effect::DoNothing => Outcome::if_running(Outcome::DoNothing, Outcome::Start),
+            },
+            State::Running => match effect {
+                Effect::Exit => {
+                    Outcome::if_running(Outcome::both(Outcome::Stop, Outcome::Exit), Outcome::Exit)
+                }
+                Effect::ProcessFailed => Outcome::if_running(Outcome::Stop, Outcome::DoNothing),
+                Effect::ProcessCompleted => Outcome::DoNothing,
+                Effect::ManifestChange => {
+                    if !self.skip_build {
+                        *state = State::Building;
+                    }
+                    self.restart_outcome(false)
+                }
+                Effect::SourceChange => {
+                    if !self.skip_build {
+                        *state = State::Building;
+                        return self.restart_outcome(false);
+                    }
+                    Outcome::DoNothing
+                }
+                Effect::ArtifactChange => self.restart_outcome(false),
+                Effect::DoNothing => Outcome::if_running(Outcome::DoNothing, Outcome::Start),
+            },
+        };
```

Comment:
This might be a little more readable written as e.g.:

```suggestion
        let outcome = match (*state, effect) {
            (_, Effect::Exit) => {
                Outcome::if_running(Outcome::both(Outcome::Stop, Outcome::Exit), Outcome::Exit)
            },
            (_, Effect::ProcessFailed) => Outcome::if_running(Outcome::Stop, Outcome::DoNothing),
            (State::Building, Effect::ProcessCompleted) => {
                *state = State::Running;
                self.restart_outcome(true)
            },
            (State::Building, Effect::ProcessCompleted) => Outcome::DoNothing,
            ...
        };
```